add holo command to generate a holography view

mattieudv opened the pull request:
Adding a feature to generate a view specifically useful for holography observations.

Do you have a way of testing this in the lab before rolling it out to site?

bmerry commented on katsdpdisp/html/client.js:
```diff
@@ -284,6 +284,9 @@ function setsignals(){
         handle_data_user_event('inputs');
+    }else if (signaltext.slice(0,4)=='holo')
+    {
+        handle_data_user_event('holo'+','+signaltext.slice(4));
```
```suggestion
        handle_data_user_event('holo,'+signaltext.slice(4));
```

bmerry commented on katsdpdisp/html/index.html:
```diff
@@ -87,6 +87,7 @@
                 <a>gain</a>
                 <a>gainwaterfall</a>
                 <a>help</a>
+				<a>holo</a>
```
Match the surrounding indentation.

bmerry commented on scripts/time_plot.py:
```diff
@@ -1497,6 +1497,69 @@ def handle_websock_event(handlerkey,*args):
                     if (websockrequest_username[thishandler]==username):
                         send_websock_cmd('ApplyViewLayout('+'["'+'","'.join([fig['figtype'] for fig in html_viewsettings[username]])+'"]'+','+str(html_layoutsettings[username]['ncols'])+')',thishandler)
+        elif (args[0].startswith('holo')):#accepts 'holo', 'holo 3' where 3 is m003 as explicit reference antenna, else first track antenna is chosen as reference antenna
+            logger.info(repr(args))
```
I haven't looked through all the code, but it seems that you're repeating this in each block, instead of just doing it once at the top? Or do you not do it for some commands?

mattieudv replied:
There is a lot of repetition but some commands have arguments and others do not. It is possible to simplify but is more effort than it is worth at this stage.

bmerry commented on scripts/time_plot.py:
```diff
@@ -1497,6 +1497,69 @@ def handle_websock_event(handlerkey,*args):
                         send_websock_cmd('ApplyViewLayout('+'["'+'","'.join([fig['figtype'] for fig in html_viewsettings[username]])+'"]'+','+str(html_layoutsettings[username]['ncols'])+')',thishandler)
+        elif (args[0].startswith('holo')):#accepts 'holo', 'holo 3' where 3 is m003 as explicit reference antenna, else first track antenna is chosen as reference antenna
+            logger.info(repr(args))
+            antnumbers=[int(antnumberstr[1:]) for antnumberstr in telstate_antenna_mask]#determine all available inputs
```
Those are antennas, not inputs.

mattieudv replied:
Well, they could also be d000 for example for CBF streams, but i guess they are not m001h.

bmerry replied:
d000 wouldn't be in telstate_antenna_mask.

mattieudv replied:
but s000 would be for ska antennas. There were cases in the past that signals like d000 existed. Whether that is the case now or not, the code that deciphers/maps signal names should be able to deal with it.

bmerry replied:
s000 is still an antenna. I'm just suggesting
```suggestion
            antnumbers=[int(antnumberstr[1:]) for antnumberstr in telstate_antenna_mask]#determine all available antennas
```

bmerry commented on scripts/time_plot.py:
```diff
@@ -1497,6 +1497,69 @@ def handle_websock_event(handlerkey,*args):
+            if (len(antnumbers)==0):
+                send_websock_cmd('logconsole("No antenna inputs found or specified",true,true,true)',handlerkey)
+            else:
+                cbid=str(telstate['sdp_capture_block_id'])
```
It's already a string - no need to `str`.

mattieudv replied:
Ok changed it to a comment. However I do not like this. Maybe you know offhand it is a string, but someone like me reading the code would not know if it is a string or a number and then wouldnt know how to handle it in the code.

mattieudv replied:
C is a much better language...

bmerry commented on scripts/time_plot.py:
```diff
@@ -1497,6 +1497,69 @@ def handle_websock_event(handlerkey,*args):
+                    track_ants=obs_params['track_ants'].split(',')
+                if 'scan_ants_always' in obs_params:
+                    scan_ants_always=obs_params['scan_ants_always'].split(',')
+                    newtrackants=[]
```
Be consistent about spacing, i.e. `new_track_ants`.

bmerry commented on scripts/time_plot.py:
```diff
@@ -1497,6 +1497,69 @@ def handle_websock_event(handlerkey,*args):
+                    scan_ants=sorted(scan_ants)
+                if (len(args)==1 or args[1]==''):
+                    if len(track_ants):
+                        refantnumber=int(track_ants[0][1:])
```
By consistent about spacing i.e. `ref_ant_number`.

bmerry commented on scripts/time_plot.py:
```diff
@@ -1497,6 +1497,69 @@ def handle_websock_event(handlerkey,*args):
+                            newtrackants.append(ant)
+                    track_ants=newtrackants
+                    scan_ants=sorted(scan_ants)
+                if (len(args)==1 or args[1]==''):
```
```suggestion
                if len(args)==1 or args[1]=='':
```

bmerry commented on scripts/time_plot.py:
```diff
@@ -1497,6 +1497,69 @@ def handle_websock_event(handlerkey,*args):
+                    else:
+                        refantnumber=antnumbers[0]
+                else:#use supplied refant
+                    refantnumberlist=parse_antennarange(','.join(args[1:]))
```
Use `ref_ant_number_list` for consistency.

mattieudv replied:
I think there are too many of these inconsistencies around, e.g. newtrackants, refantnumber, refantnumberlist, cbid, antnumbers, and also in other nearby blocks of code. It is probably better to just leave it unless a concerted effort is done to change everything throughout all the source code.

bmerry replied:
One day we're going to make you actually fix all the awfulness that you keep putting off because all the surrounding code is awful :-)

mattieudv replied:
@ludwigschwardt spotted some deprecated usages of telstate and a few other points. I told him I should do a different pull request to address a bunch of random things like that (maybe next month). So it would be good if you compile a few suggestions for such a PR next month or so.

bmerry commented on scripts/time_plot.py:
```diff
@@ -1497,6 +1497,69 @@ def handle_websock_event(handlerkey,*args):
+                else:#use supplied refant
+                    refantnumberlist=parse_antennarange(','.join(args[1:]))
+                    if (len(refantnumberlist)==1):
+                        refantnumber=refantnumberlist[0]
```
What happens if this is not an antenna that's in the subarray?

bmerry replied:
This is a question that you haven't answered.

bmerry replied:
This is still outstanding.

bmerry commented on scripts/time_plot.py:
```diff
@@ -1497,6 +1497,69 @@ def handle_websock_event(handlerkey,*args):
+                if len(scan_ants)==0:
+                    scan_ant_numbers=range(64)
+                else:
+                    scan_ant_numbers=[int(antname[1:]) for antname in scan_ants if antname!=refantnumber]
```
This will presumably be an empty list if there is exactly one scan antenna and it is also the reference antenna. Is that a problem?

bmerry commented on scripts/time_plot.py:
```diff
@@ -1497,6 +1497,69 @@ def handle_websock_event(handlerkey,*args):
+
+                for iant in scan_ant_numbers: # keep blank placeholder if antenna not present
+                    if (iant<refantnumber):
+                        ijstr=str(iant)+str(args[0][-1])+str(refantnumber)+str(args[0][-2])
```
Could be simplified with
```suggestion
                        ijstr=f'{iant}{args[0][-1]}{refantnumber}{args[0][-2]}'
```

bmerry commented on scripts/time_plot.py:
```diff
@@ -1497,6 +1497,69 @@ def handle_websock_event(handlerkey,*args):
+                    scan_ant_numbers=[int(antname[1:]) for antname in scan_ants if antname!=refantnumber]
+
+                for iant in scan_ant_numbers: # keep blank placeholder if antenna not present
+                    if (iant<refantnumber):
```
```suggestion
                    if iant<refantnumber:
```

bmerry commented on scripts/time_plot.py:
```diff
@@ -1497,6 +1497,69 @@ def handle_websock_event(handlerkey,*args):
+                    if (iant<refantnumber):
+                        ijstr=str(iant)+str(args[0][-1])+str(refantnumber)+str(args[0][-2])
+                    else:
+                        ijstr=str(refantnumber)+str(args[0][-2])+str(iant)+str(args[0][-1])
```
Are `args[0][-2]` and `args[0][-1]` supposed to be polarisations? None of the documentation mentions anything about specifying polarisations.

mattieudv replied:
yes, i guess that would be a bug

bmerry commented on scripts/time_plot.py:
```diff
@@ -1497,6 +1497,69 @@ def handle_websock_event(handlerkey,*args):
+                    decodedsignal=decodecustomsignal(ijstr)
+                    html_customsignals[username].append(decodedsignal)
+
+                html_viewsettings[username].append({'figtype':'timeseries','type':'pow','xtype':'s'  ,'xmin':[],'xmax':[],'ymin':[],'ymax':[],'cmin':[],'cmax':[],'showlegend':'on','showxlabel':'off','showylabel':'off','showxticklabel':'on','showyticklabel':'on','showtitle':'on','processtime':0,'version':0,'sensor':'m%03d_pos_actual_scan_elev'%refantnumber})
```
```suggestion
                html_viewsettings[username].append({'figtype':'timeseries','type':'pow','xtype':'s','xmin':[],'xmax':[],'ymin':[],'ymax':[],'cmin':[],'cmax':[],'showlegend':'on','showxlabel':'off','showylabel':'off','showxticklabel':'on','showyticklabel':'on','showtitle':'on','processtime':0,'version':0,'sensor':'m%03d_pos_actual_scan_elev'%refantnumber})
```

bmerry commented on scripts/time_plot.py:
```diff
@@ -1497,6 +1497,69 @@ def handle_websock_event(handlerkey,*args):
+                    html_customsignals[username].append(decodedsignal)
+
+                html_viewsettings[username].append({'figtype':'timeseries','type':'pow','xtype':'s'  ,'xmin':[],'xmax':[],'ymin':[],'ymax':[],'cmin':[],'cmax':[],'showlegend':'on','showxlabel':'off','showylabel':'off','showxticklabel':'on','showyticklabel':'on','showtitle':'on','processtime':0,'version':0,'sensor':'m%03d_pos_actual_scan_elev'%refantnumber})
+                html_viewsettings[username].append({'figtype':'timeseries','type':'pow','xtype':'s'  ,'xmin':[],'xmax':[],'ymin':[],'ymax':[],'cmin':[],'cmax':[],'showlegend':'on','showxlabel':'off','showylabel':'off','showxticklabel':'on','showyticklabel':'on','showtitle':'on','processtime':0,'version':0,'sensor':'m%03d_pos_actual_scan_elev'%(scan_ant_numbers[0])})
```
```suggestion
                html_viewsettings[username].append({'figtype':'timeseries','type':'pow','xtype':'s','xmin':[],'xmax':[],'ymin':[],'ymax':[],'cmin':[],'cmax':[],'showlegend':'on','showxlabel':'off','showylabel':'off','showxticklabel':'on','showyticklabel':'on','showtitle':'on','processtime':0,'version':0,'sensor':'m%03d_pos_actual_scan_elev'%(scan_ant_numbers[0])})
```

bmerry commented on katsdpdisp/html/client.js:
```diff
@@ -286,7 +286,10 @@ function setsignals(){
     {
-        handle_data_user_event('holo'+','+signaltext.slice(4));
+        handle_data_user_event('holohh,'+signaltext.slice(4));
+    }else if (signaltext.slice(0,6)=='holohh' || signaltext.slice(0,6)=='holovv' || signaltext.slice(0,6)=='holovh' || signaltext.slice(0,6)=='holovh')
```
You've got holovh twice and no holohv.

bmerry commented on katsdpdisp/html/client.js:
```diff
@@ -286,7 +286,10 @@ function setsignals(){
+        handle_data_user_event('holohh,'+signaltext.slice(4));
+    }else if (signaltext.slice(0,6)=='holohh' || signaltext.slice(0,6)=='holovv' || signaltext.slice(0,6)=='holovh' || signaltext.slice(0,6)=='holovh')
+    {
+        handle_data_user_event(signaltext.slice(0,6)+','+signaltext.slice(6));
```
This code will never be reached because it will match the `signaltext.slice(0,4)=='holo'` test above.

bmerry commented on scripts/time_plot.py:
```diff
@@ -1497,6 +1497,69 @@ def handle_websock_event(handlerkey,*args):
                 for thishandler in websockrequest_username.keys():
                     if (websockrequest_username[thishandler]==username):
                         send_websock_cmd('ApplyViewLayout('+'["'+'","'.join([fig['figtype'] for fig in html_viewsettings[username]])+'"]'+','+str(html_layoutsettings[username]['ncols'])+')',thishandler)
+        elif (args[0].startswith('holo')):#accepts 'holohh', 'holovv', 'holohv', 'holovh', 'holo 3' where 3 is m003 as explicit reference antenna, else first track antenna is chosen as reference antenna
```
'holo 3' is presumably no longer valid here, since it gets rewritten to holohh?

mattieudv replied:
The comment is a bit misleading, from user side it should still work because holo is replaced with holohh in client.js

mattieudv replied:
client.js should separate the 3 as args[1]

bmerry replied:
But what's received here will be 'holohh 3', right?

mattieudv replied:
I am expecting args=['holohh','3']

bmerry replied:
```suggestion
        elif (args[0].startswith('holo')):#accepts 'holohh', 'holovv', 'holohv', 'holovh', 'holohh 3' where 3 is m003 as explicit reference antenna, else first track antenna is chosen as reference antenna
```

mattieudv commented on scripts/time_plot.py:
```diff
@@ -3229,23 +3292,25 @@ def pack_binarydata_msg(varname,val,dtype):
 
 #Caught exception (local variable 'action' referenced before assignment). Removing registered handler
-def parse_websock_cmd(s, request):
+def parse_websock_cmd(s, handlerkey):
```
@bmerry just do a final squiz from here onwards as these changes have widespread affects whereas the changes above are only relevant to the issuing of the 'holo' command.

bmerry commented on katsdpdisp/html/help.txt:
```diff
@@ -232,8 +232,9 @@ A spectrum plot is also produced.
 For example:
-'holo' regenerates a view using the first tracking antenna as reference of cross correlation product
+'holo' regenerates a view using the first tracking antenna as reference of cross correlation product (h pol)
 'holo 4' regenerates a view using m004 as reference of cross correlation product
```
Is `holo 4` also implicitly h pol?

mattieudv replied:
yes

bmerry replied:
Help should say so then.

bmerry commented on scripts/time_plot.py:
```diff
@@ -1497,23 +1498,23 @@ def handle_websock_event(handlerkey,*args):
+                if b'track_ants' in obs_params:
+                    track_ants=to_str(obs_params[b'track_ants']).split(',')
+                if b'scan_ants_always' in obs_params:
+                    scan_ants_always=to_str(obs_params[b'scan_ants_always']).split(',')
```
Eventually we'll get CAM to start using Python 3 and then obs_params will use strings. Instead of changing all this code to use bytes, you can just change how you get obs_params to `obs_params = to_str(telstate.get(obs_params_keys, {}))`, which will convert all the keys and values to strings.

mattieudv replied:
yes, they should all be changed in a future PR, to to_str

bmerry replied:
Since this is the first time you're using to_str, you cshould start using it right in this PR i.e. use `to_str` when you get obs_params out of telstate instead of on each access to it.

bmerry replied:
To be clear, I'm not suggesting you should try to roll out to_str to the rest of the code - just update the code that's new in this PR.

bmerry reviewed (changes requested):
Still a few questions open from my first review.

mattieudv commented:
> Still a few questions open from my first review.

I do not see these questions - most code blocks are outdated.

I added 'print scan_ants_always to console', would ideally like to merge this branch to master now for today's deployment.

bmerry commented on scripts/time_plot.py:
```diff
@@ -1497,6 +1498,73 @@ def handle_websock_event(handlerkey,*args):
+            else:
+                cbid=telstate['sdp_capture_block_id']#this is a string
+                obs_params_key=telstate.join(cbid, 'obs_params')
+                obs_params=telstate.get(obs_params_key, {})
+                scan_ants_always=[]
+                scan_ants=[]
+                track_ants=[]
+                if b'scan_ants' in obs_params:
+                    scan_ants=to_str(obs_params[b'scan_ants']).split(',')
+                if b'track_ants' in obs_params:
+                    track_ants=to_str(obs_params[b'track_ants']).split(',')
+                if b'scan_ants_always' in obs_params:
+                    scan_ants_always=to_str(obs_params[b'scan_ants_always']).split(',')
```
```suggestion
                obs_params=to_str(telstate.get(obs_params_key, {}))
                scan_ants_always=[]
                scan_ants=[]
                track_ants=[]
                if 'scan_ants' in obs_params:
                    scan_ants=obs_params['scan_ants'].split(',')
                if 'track_ants' in obs_params:
                    track_ants=obs_params['track_ants'].split(',')
                if 'scan_ants_always' in obs_params:
                    scan_ants_always=obs_params['scan_ants_always'].split(',')
```

mattieudv commented:
Anything outstanding about this, can we merge with master?

On Wed, Apr 21, 2021 at 10:23 AM Bruce Merry ***@***.***>
wrote:

> ***@***.**** requested changes on this pull request.
> ------------------------------
>
> In scripts/time_plot.py
> <https://github.com/ska-sa/katsdpdisp/pull/84#discussion_r617306010>:
>
> > +                obs_params=telstate.get(obs_params_key, {})
>
> +                scan_ants_always=[]
>
> +                scan_ants=[]
>
> +                track_ants=[]
>
> +                if b'scan_ants' in obs_params:
>
> +                    scan_ants=to_str(obs_params[b'scan_ants']).split(',')
>
> +                if b'track_ants' in obs_params:
>
> +                    track_ants=to_str(obs_params[b'track_ants']).split(',')
>
> +                if b'scan_ants_always' in obs_params:
>
> +                    scan_ants_always=to_str(obs_params[b'scan_ants_always']).split(',')
>
>
> ⬇️ Suggested change
>
> -                obs_params=telstate.get(obs_params_key, {})
>
> -                scan_ants_always=[]
>
> -                scan_ants=[]
>
> -                track_ants=[]
>
> -                if b'scan_ants' in obs_params:
>
> -                    scan_ants=to_str(obs_params[b'scan_ants']).split(',')
>
> -                if b'track_ants' in obs_params:
>
> -                    track_ants=to_str(obs_params[b'track_ants']).split(',')
>
> -                if b'scan_ants_always' in obs_params:
>
> -                    scan_ants_always=to_str(obs_params[b'scan_ants_always']).split(',')
>
> +                obs_params=to_str(telstate.get(obs_params_key, {}))
>
> +                scan_ants_always=[]
>
> +                scan_ants=[]
>
> +                track_ants=[]
>
> +                if 'scan_ants' in obs_params:
>
> +                    scan_ants=obs_params['scan_ants'].split(',')
>
> +                if 'track_ants' in obs_params:
>
> +                    track_ants=obs_params['track_ants'].split(',')
>
> +                if 'scan_ants_always' in obs_params:
>
> +                    scan_ants_always=obs_params['scan_ants_always'].split(',')
>
>
> ------------------------------
>
> In scripts/time_plot.py
> <https://github.com/ska-sa/katsdpdisp/pull/84#discussion_r617306219>:
>
> > +                    for ant in track_ants:
>
> +                        if ant in scan_ants_always:
>
> +                            scan_ants.append(ant)
>
> +                        else:
>
> +                            newtrackants.append(ant)
>
> +                    track_ants=newtrackants
>
> +                    scan_ants=sorted(scan_ants)
>
> +                if (len(args)==1 or args[1]==''):
>
> +                    if len(track_ants):
>
> +                        refantnumber=int(track_ants[0][1:])
>
> +                    else:
>
> +                        refantnumber=antnumbers[0]
>
> +                else:#use supplied refant
>
> +                    refantnumberlist=parse_antennarange(','.join(args[1:]))
>
> +                    if (len(refantnumberlist)==1):
>
> +                        refantnumber=refantnumberlist[0]
>
>
> This is a question that you haven't answered.
> ------------------------------
>
> In katsdpdisp/html/help.txt
> <https://github.com/ska-sa/katsdpdisp/pull/84#discussion_r617302388>:
>
> > @@ -232,8 +232,9 @@ A spectrum plot is also produced.
>
>  You may additionally want to issue a command such as 'flags=..500,520..'
>
>
>
>  For example:
>
> -'holo' regenerates a view using the first tracking antenna as reference of cross correlation product
>
> +'holo' regenerates a view using the first tracking antenna as reference of cross correlation product (h pol)
>
>  'holo 4' regenerates a view using m004 as reference of cross correlation product
>
>
> Help should say so then.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ska-sa/katsdpdisp/pull/84#pullrequestreview-640755219>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABKRCLEH5MURJH7ZZFUZKLTTJ2DOLANCNFSM4ZLHZQTA>
> .
>

mattieudv commented:
github appears to suck, there is only a resolve option, and no reply option
to your comment.

In the case you mention the 'else' part will be called of the if statement
(i.e. it simply chooses the first antenna in the array):
                        send_websock_cmd('logconsole("Invalid reference
antenna specified, using default instead",true,true,true)',handlerkey)
                        refantnumber=antnumbers[0]


On Tue, May 4, 2021 at 4:22 PM Bruce Merry ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In scripts/time_plot.py
> <https://github.com/ska-sa/katsdpdisp/pull/84#discussion_r625824631>:
>
> > +                    for ant in track_ants:
> +                        if ant in scan_ants_always:
> +                            scan_ants.append(ant)
> +                        else:
> +                            newtrackants.append(ant)
> +                    track_ants=newtrackants
> +                    scan_ants=sorted(scan_ants)
> +                if (len(args)==1 or args[1]==''):
> +                    if len(track_ants):
> +                        refantnumber=int(track_ants[0][1:])
> +                    else:
> +                        refantnumber=antnumbers[0]
> +                else:#use supplied refant
> +                    refantnumberlist=parse_antennarange(','.join(args[1:]))
> +                    if (len(refantnumberlist)==1):
> +                        refantnumber=refantnumberlist[0]
>
> This is still outstanding.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ska-sa/katsdpdisp/pull/84#discussion_r625824631>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABKRCLAVIYMGSVFZWU237QLTL77ITANCNFSM4ZLHZQTA>
> .
>

bmerry reviewed (approved):
Hooray, reduced code and it's more robust :-)

mattieudv commented:
Merging now
